### PR TITLE
Roller

### PR DESCRIPTION
**Pull request by danrs**
R2-D7 Can now roll dice with the `!roll` command! Features are as follows:

- Simulated dice rolls, including modifications
- Roll just one colour of dice, or roll reds against greens
- User-friendly regex parsing makes the command fairly intuitive
- Informative syntax help command
- Integration with [punkuser's calculator](http://gateofstorms.net/2/multi/) for expected # hits
- Unit tests to make freakydug happy

### CODE REVIEW

File: r2d7/roller.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+class ModdedRoll(object):
+    """
+    Class for managing all aspects of a roll. It parses, it rolls, it queries the online calculator.
+    """
+    _re_def = '\\b((green)|(g))\\b'
+    _re_atk = '\\b((red)|(r))\\b'
+    pattern_def = re.compile(_re_def, re.I)
+    pattern_atk = re.compile(_re_atk, re.I)
+    pattern_main = re.compile('(?P<num_dice>[0-9]+) ?(?P<color>%s|%s)' % (_re_def, _re_atk), re.I)
```
Comment (FreakyDug) [inline r2d7/roller.py]:
```suggestion
    pattern_main = re.compile(f'(?P<num_dice>[0-9]+) ?(?P<color>{_re_def}|{_re_atk})', re.I)
```

File: r2d7/roller.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            raise RollSyntaxError('I don\'t know how many dice you want me to roll')
+
+        if self.num_dice > 100:
+            logger.debug('Too many dice requested (max: 100, requested: %d)' % (self.num_dice))
+            raise RollSyntaxError('Sorry, I can\'t carry more than 100 dice, and chopper won\'t help :chopper:')
+        elif self.num_dice < 1:
+            logger.debug('Too few dice requested (min: 1, requested: %d)' % (self.num_dice))
+            raise RollSyntaxError('Sorry, I can\'t roll fewer than 1 dice :jar_jar:')
+
+        color_string = match_main.group('color') or ''
```
Comment (FreakyDug) [inline r2d7/roller.py]: This doesn't appear to be used.

File: r2d7/roller.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+        if self.reroll > 3:
+            logger.debug('Too many rerolls requested (max: 3, requested: %d)' % (self.reroll))
+            raise RollSyntaxError('Sorry, the calculator only allows up to 3 rerolls')
+
+        self.modify_dice()
+
+    def parse_mod_boolean(self, attr_name, message):
+        value = ModdedRoll.pattern_mod[attr_name].search(message) is not None
+        object.__setattr__(self, attr_name, value)
```
Comment (FreakyDug) [inline r2d7/roller.py]: Can't you just use `setattr` here?

File: r2d7/roller.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        match_mod = ModdedRoll.pattern_mod[attr_name].search(message)
+        if match_mod:
+            match_count = ModdedRoll.pattern_count.search(match_mod.group(1))
+            value = int(match_count.group(1)) if match_count is not None else 1
+        else:
+            value = 0
+        object.__setattr__(self, attr_name, value)
+
+    def modify_dice(self):
+        if len(self.dice) == 0:
```
Comment (FreakyDug) [inline r2d7/roller.py]:
```suggestion
        if not self.dice:
```

File: r2d7/roller.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                force = force - 1 if focussed else force
+            if not focussed and evades > 0 and d.die_type == DieType.defense:
+                evades = evades - 1 if d.evade() else evades
+
+    def roll_safe(self):
+        return len(self.dice) <= CalculatorForm.max_dice and self.reinforce <= CalculatorForm.max_reinforce
+
+    def calculator_form(self):
+        if self.die_type == DieType.attack:
+            form = AttackForm(dice = len(self.dice),
```
Comment (FreakyDug) [inline r2d7/roller.py]: I suggest putting a newline after `(` so all the kwargs are indented the same amount.

File: r2d7/roller.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    def actual_roll(self):
+        output = ''.join([str(d) for d in self.dice])
+        # reinforce is a special case
+        # It can't be meaningfully added without knowing the attack roll
+        # Instead of implementing it as an added result, just show that
+        # there were reinforce tokens in play. This is just as useful to
+        # users and is much simpler to implement.
+        if self.reinforce > 0 and self.die_type == DieType.defense:
+            output = output + ' + ' + ':reinforce:'*self.reinforce
```
Comment (FreakyDug) [inline r2d7/roller.py]:
```suggestion
            output = f"{output} + {':reinforce:'*self.reinforce}"
```

File: r2d7/roller.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        return output
+
+    def output(self):
+        output = []
+        output.append('You rolled: %s' % self.actual_roll())
```
Comment (FreakyDug) [inline r2d7/roller.py]:
```suggestion
        output.append(f'You rolled: {self.actual_roll()}')
```

**Comment by danrs** [comment]
All good suggestions! You're right about f-strings - I always forget the syntax for them but they are much nicer

**Comment by danrs** [comment]
Merge when ready, captain!

**Comment by FreakyDug** [comment]
I've sent you an invite to be a collaborator so you can merge stuff yourself. I think it makes sense for you to merge it when you're ready as you can fix it more easily if it breaks.

**Pull request merged**